[FIX] PDF download routes — correct renderToBuffer implementation

@claude

## Overview
The PDF download routes have been failing TypeScript checks due to
incorrect usage of `renderToBuffer` from `@react-pdf/renderer`.
This ticket provides the exact correct implementation to use.

Do NOT deviate from the code below. Do not use JSX syntax in these
route files. Do not import from `@/lib/supabase/server` — the correct
import is `@/lib/supabase`.

---

## File 1: `app/api/download-pdf/resume/route.ts`
Replace the entire file with exactly this:
```typescript
import { NextRequest, NextResponse } from 'next/server'
import { auth, currentUser } from '@clerk/nextjs/server'
import { renderToBuffer } from '@react-pdf/renderer'
import { createElement } from 'react'
import { supabaseServer } from '@/lib/supabase'
import ResumePDF from '@/lib/pdf/ResumePDF'

export const runtime = 'nodejs'

export async function POST(request: NextRequest) {
  try {
    const { userId } = await auth()
    if (!userId) {
      return NextResponse.json({ error: 'Unauthorized' }, { status: 401 })
    }

    const { applicationId } = await request.json()
    if (!applicationId) {
      return NextResponse.json({ error: 'applicationId is required' }, { status: 400 })
    }

    const supabase = await supabaseServer()

    const { data: application, error } = await supabase
      .from('applications')
      .select('*')
      .eq('id', applicationId)
      .single()

    if (error || !application) {
      return NextResponse.json({ error: 'Application not found' }, { status: 404 })
    }

    if (application.user_id !== userId) {
      return NextResponse.json({ error: 'Forbidden' }, { status: 403 })
    }

    if (!application.generated_resume) {
      return NextResponse.json({ error: 'No resume content found' }, { status: 404 })
    }

    const user = await currentUser()
    const fullName = user?.fullName || user?.firstName || 'User'
    const lastName = user?.lastName || fullName.split(' ').pop() || 'User'

    const props = {
      resumeText: application.generated_resume,
      candidateName: fullName,
      company: application.company,
      jobTitle: application.job_title,
    }

    const element = createElement(ResumePDF, props)
    const pdfBuffer = await renderToBuffer(element as React.ReactElement<any>)

    const companyName = application.company.replace(/\s+/g, '_')
    const role = application.job_title.replace(/\s+/g, '_')
    const filename = `${lastName}_Resume_${companyName}_${role}.pdf`

    return new NextResponse(new Uint8Array(pdfBuffer), {
      headers: {
        'Content-Type': 'application/pdf',
        'Content-Disposition': `attachment; filename="${filename}"`,
      },
    })
  } catch (error) {
    console.error('PDF generation error:', error)
    return NextResponse.json({ error: 'Failed to generate PDF' }, { status: 500 })
  }
}
```

---

## File 2: `app/api/download-pdf/cover-letter/route.ts`
Replace the entire file with exactly this:
```typescript
import { NextRequest, NextResponse } from 'next/server'
import { auth, currentUser } from '@clerk/nextjs/server'
import { renderToBuffer } from '@react-pdf/renderer'
import { createElement } from 'react'
import { supabaseServer } from '@/lib/supabase'
import CoverLetterPDF from '@/lib/pdf/CoverLetterPDF'

export const runtime = 'nodejs'

export async function POST(request: NextRequest) {
  try {
    const { userId } = await auth()
    if (!userId) {
      return NextResponse.json({ error: 'Unauthorized' }, { status: 401 })
    }

    const { applicationId } = await request.json()
    if (!applicationId) {
      return NextResponse.json({ error: 'applicationId is required' }, { status: 400 })
    }

    const supabase = await supabaseServer()

    const { data: application, error } = await supabase
      .from('applications')
      .select('*')
      .eq('id', applicationId)
      .single()

    if (error || !application) {
      return NextResponse.json({ error: 'Application not found' }, { status: 404 })
    }

    if (application.user_id !== userId) {
      return NextResponse.json({ error: 'Forbidden' }, { status: 403 })
    }

    if (!application.generated_cover_letter) {
      return NextResponse.json({ error: 'No cover letter content found' }, { status: 404 })
    }

    const user = await currentUser()
    const fullName = user?.fullName || user?.firstName || 'User'
    const lastName = user?.lastName || fullName.split(' ').pop() || 'User'

    const props = {
      coverLetterText: application.generated_cover_letter,
      candidateName: fullName,
      company: application.company,
      jobTitle: application.job_title,
    }

    const element = createElement(CoverLetterPDF, props)
    const pdfBuffer = await renderToBuffer(element as React.ReactElement<any>)

    const companyName = application.company.replace(/\s+/g, '_')
    const role = application.job_title.replace(/\s+/g, '_')
    const filename = `${lastName}_CoverLetter_${companyName}_${role}.pdf`

    return new NextResponse(new Uint8Array(pdfBuffer), {
      headers: {
        'Content-Type': 'application/pdf',
        'Content-Disposition': `attachment; filename="${filename}"`,
      },
    })
  } catch (error) {
    console.error('PDF generation error:', error)
    return NextResponse.json({ error: 'Failed to generate PDF' }, { status: 500 })
  }
}
```

---

## Why these exact patterns
- `createElement` instead of JSX — avoids parser issues in nodejs API routes
- `as React.ReactElement<any>` cast — the PDF components already wrap
  themselves in `<Document>` so the DocumentProps type mismatch is a
  false positive
- `new Uint8Array(pdfBuffer)` — correct way to pass Buffer as Response body
- `supabaseServer` from `@/lib/supabase` — matches the actual file in
  this repo, not `@/lib/supabase/server` which does not exist
- `generated_resume` / `generated_cover_letter` — matches actual column
  names in the applications table

## Definition of Done
- Both files match the code above exactly
- `npx tsc --noEmit` passes with zero errors
- No imports from `@/lib/supabase/server`
- No JSX syntax in either route file

## ❌ TypeScript Check Failed — Fix Required (Attempt 1/3)

@claude Your changes have TypeScript errors. You cannot run
`tsc` directly — read the errors below and fix them in the
source files, then push again.

```
app/api/download-pdf/cover-letter/route.ts(38,22): error TS2339: Property 'generated_cover_letter' does not exist on type '{ id: string; user_id: string; resume_id: string | null; job_title: string; company: string; job_description: string; resume_content: string | null; cover_letter_content: string | null; status: "applied" | ... 3 more ... | "withdrawn"; application_date: string; created_at: string; updated_at: string; }'.
app/api/download-pdf/cover-letter/route.ts(47,36): error TS2339: Property 'generated_cover_letter' does not exist on type '{ id: string; user_id: string; resume_id: string | null; job_title: string; company: string; job_description: string; resume_content: string | null; cover_letter_content: string | null; status: "applied" | ... 3 more ... | "withdrawn"; application_date: string; created_at: string; updated_at: string; }'.
app/api/download-pdf/resume/route.ts(38,22): error TS2339: Property 'generated_resume' does not exist on type '{ id: string; user_id: string; resume_id: string | null; job_title: string; company: string; job_description: string; resume_content: string | null; cover_letter_content: string | null; status: "applied" | ... 3 more ... | "withdrawn"; application_date: string; created_at: string; updated_at: string; }'.
app/api/download-pdf/resume/route.ts(47,31): error TS2339: Property 'generated_resume' does not exist on type '{ id: string; user_id: string; resume_id: string | null; job_title: string; company: string; job_description: string; resume_content: string | null; cover_letter_content: string | null; status: "applied" | ... 3 more ... | "withdrawn"; application_date: string; created_at: string; updated_at: string; }'.
```

**Before fixing — read TYPES.md first.**

**Most common causes in this repo:**
- String not in union type → add it to the type definition first
- Cannot redeclare variable → use a unique variable name
- Cannot find name → declare the variable in the current scope
- chunk.delta.text → narrow with chunk.delta.type === "text_delta" first

Fix ALL errors above, not just the first one.

--- a/app/api/download-pdf/cover-letter/route.ts
+++ b/app/api/download-pdf/cover-letter/route.ts
@@ -1,78 +1,70 @@
-export const runtime = 'nodejs';
+import { NextRequest, NextResponse } from 'next/server'
+import { auth, currentUser } from '@clerk/nextjs/server'
+import { renderToBuffer } from '@react-pdf/renderer'
+import { createElement } from 'react'
+import { supabaseServer } from '@/lib/supabase'
+import CoverLetterPDF from '@/lib/pdf/CoverLetterPDF'
 
-import { auth } from '@clerk/nextjs/server';
-import { NextRequest } from 'next/server';
-import { supabaseServer } from '@/lib/supabase';
-import React from 'react';
-import { renderToBuffer } from '@react-pdf/renderer';
-import CoverLetterPDF from '@/lib/pdf/CoverLetterPDF';
+export const runtime = 'nodejs'
 
-export async function POST(req: NextRequest) {
+export async function POST(request: NextRequest) {
   try {
-    // Verify user is signed in
-    const { userId } = await auth();
+    const { userId } = await auth()
     if (!userId) {
-      return new Response('Unauthorized', { status: 401 });
+      return NextResponse.json({ error: 'Unauthorized' }, { status: 401 })
     }
 
-    // Parse request body
-    const { applicationId } = await req.json();
-    
+    const { applicationId } = await request.json()
     if (!applicationId) {
-      return new Response('Missing applicationId', { status: 400 });
+      return NextResponse.json({ error: 'applicationId is required' }, { status: 400 })
     }
 
-    // Get application from Supabase
-    const supabase = await supabaseServer();
+    const supabase = await supabaseServer()
+
     const { data: application, error } = await supabase
       .from('applications')
       .select('*')
       .eq('id', applicationId)
-      .single();
+      .single()
 
-    if (error) {
-      console.error('Error fetching application:', error);
-      return new Response('Application not found', { status: 404 });
+    if (error || !application) {
+      return NextResponse.json({ error: 'Application not found' }, { status: 404 })
     }
 
-    // Verify application belongs to current user
     if (application.user_id !== userId) {
-      return new Response('Forbidden', { status: 403 });
+      return NextResponse.json({ error: 'Forbidden' }, { status: 403 })
     }
 
-    // Check if cover letter content exists
     if (!application.cover_letter_content) {
-      return new Response('Cover letter content not found', { status: 404 });
+      return NextResponse.json({ error: 'No cover letter content found' }, { status: 404 })
     }
 
-    // Get user name from Clerk (fallback approach)
-    const candidateName = 'User';
+    const user = await currentUser()
+    const fullName = user?.fullName || user?.firstName || 'User'
+    const lastName = user?.lastName || fullName.split(' ').pop() || 'User'
 
-    // Generate PDF using CoverLetterPDF component
-    const pdfElement = React.createElement(CoverLetterPDF, {
+    const props = {
       coverLetterText: application.cover_letter_content,
-      candidateName: candidateName,
+      candidateName: fullName,
       company: application.company,
       jobTitle: application.job_title,
-    });
+    }
 
-    const pdfBuffer = await renderToBuffer(pdfElement);
+    const element = createElement(CoverLetterPDF, props)
+    const pdfBuffer = await renderToBuffer(element as React.ReactElement<any>)
 
-    // Create filename with proper format: LastName_CoverLetter_Company_Role.pdf
-    const safeCompany = application.company.replace(/[^a-zA-Z0-9]/g, '_');
-    const safeRole = application.job_title.replace(/[^a-zA-Z0-9]/g, '_');
-    const filename = `CoverLetter_${safeCompany}_${safeRole}.pdf`;
+    const companyName = application.company.replace(/\s+/g, '_')
+    const role = application.job_title.replace(/\s+/g, '_')
+    const filename = `${lastName}_CoverLetter_${companyName}_${role}.pdf`
 
-    // Return PDF as download
-    return new Response(new Uint8Array(pdfBuffer), {
+    return new NextResponse(new Uint8Array(pdfBuffer), {
       headers: {
         'Content-Type': 'application/pdf',
         'Content-Disposition': `attachment; filename="${filename}"`,
       },
-    });
-
+    })
   } catch (error) {
-    console.error('PDF generation error:', error);
-    return new Response('Internal server error', { status: 500 });
+    console.error('PDF generation error:', error)
+    return NextResponse.json({ error: 'Failed to generate PDF' }, { status: 500 })
   }
 }

--- a/app/api/download-pdf/resume/route.ts
+++ b/app/api/download-pdf/resume/route.ts
@@ -1,78 +1,70 @@
-export const runtime = 'nodejs';
+import { NextRequest, NextResponse } from 'next/server'
+import { auth, currentUser } from '@clerk/nextjs/server'
+import { renderToBuffer } from '@react-pdf/renderer'
+import { createElement } from 'react'
+import { supabaseServer } from '@/lib/supabase'
+import ResumePDF from '@/lib/pdf/ResumePDF'
 
-import { auth } from '@clerk/nextjs/server';
-import { NextRequest } from 'next/server';
-import { supabaseServer } from '@/lib/supabase';
-import React from 'react';
-import { renderToBuffer } from '@react-pdf/renderer';
-import ResumePDF from '@/lib/pdf/ResumePDF';
+export const runtime = 'nodejs'
 
-export async function POST(req: NextRequest) {
+export async function POST(request: NextRequest) {
   try {
-    // Verify user is signed in
-    const { userId } = await auth();
+    const { userId } = await auth()
     if (!userId) {
-      return new Response('Unauthorized', { status: 401 });
+      return NextResponse.json({ error: 'Unauthorized' }, { status: 401 })
     }
 
-    // Parse request body
-    const { applicationId } = await req.json();
-    
+    const { applicationId } = await request.json()
     if (!applicationId) {
-      return new Response('Missing applicationId', { status: 400 });
+      return NextResponse.json({ error: 'applicationId is required' }, { status: 400 })
     }
 
-    // Get application from Supabase
-    const supabase = await supabaseServer();
+    const supabase = await supabaseServer()
+
     const { data: application, error } = await supabase
       .from('applications')
       .select('*')
       .eq('id', applicationId)
-      .single();
+      .single()
 
-    if (error) {
-      console.error('Error fetching application:', error);
-      return new Response('Application not found', { status: 404 });
+    if (error || !application) {
+      return NextResponse.json({ error: 'Application not found' }, { status: 404 })
     }
 
-    // Verify application belongs to current user
     if (application.user_id !== userId) {
-      return new Response('Forbidden', { status: 403 });
+      return NextResponse.json({ error: 'Forbidden' }, { status: 403 })
     }
 
-    // Check if resume content exists
     if (!application.resume_content) {
-      return new Response('Resume content not found', { status: 404 });
+      return NextResponse.json({ error: 'No resume content found' }, { status: 404 })
     }
 
-    // Get user name from Clerk (fallback approach)
-    const candidateName = 'Resume';
+    const user = await currentUser()
+    const fullName = user?.fullName || user?.firstName || 'User'
+    const lastName = user?.lastName || fullName.split(' ').pop() || 'User'
 
-    // Generate PDF using ResumePDF component
-    const pdfElement = React.createElement(ResumePDF, {
+    const props = {
       resumeText: application.resume_content,
-      candidateName: candidateName,
+      candidateName: fullName,
       company: application.company,
       jobTitle: application.job_title,
-    });
+    }
 
-    const pdfBuffer = await renderToBuffer(pdfElement);
+    const element = createElement(ResumePDF, props)
+    const pdfBuffer = await renderToBuffer(element as React.ReactElement<any>)
 
-    // Create filename with proper format: LastName_Resume_Company_Role.pdf
-    const safeCompany = application.company.replace(/[^a-zA-Z0-9]/g, '_');
-    const safeRole = application.job_title.replace(/[^a-zA-Z0-9]/g, '_');
-    const filename = `Resume_${safeCompany}_${safeRole}.pdf`;
+    const companyName = application.company.replace(/\s+/g, '_')
+    const role = application.job_title.replace(/\s+/g, '_')
+    const filename = `${lastName}_Resume_${companyName}_${role}.pdf`
 
-    // Return PDF as download
-    return new Response(new Uint8Array(pdfBuffer), {
+    return new NextResponse(new Uint8Array(pdfBuffer), {
       headers: {
         'Content-Type': 'application/pdf',
         'Content-Disposition': `attachment; filename="${filename}"`,
       },
-    });
-
+    })
   } catch (error) {
-    console.error('PDF generation error:', error);
-    return new Response('Internal server error', { status: 500 });
+    console.error('PDF generation error:', error)
+    return NextResponse.json({ error: 'Failed to generate PDF' }, { status: 500 })
   }
 }

